openedx/features/qverse_features/registration/signals.py

```diff
@@ -1,20 +1,172 @@
+        return
+    dict_reader = DictReader(csv_file, delimiter=',')
+    # The header of uploaded csv might be in uppercase, lowercase or in camel case
+    # so to make our code independent to case just turn all the keys of reader into lower case.
+    # Also in python standard library for csv there is an option named skipinitialwhitespaces
+    # which is only applied for leading whitespaces and there is no option for trailing
+    # whitespaces. So, we will have to handle that case ourselves
+    reader = (dict((k.strip().lower(), v.strip()) for k, v in row.items()) for row in dict_reader)
+    output_file_rows = []
+    for row in reader:
+        try:
+            output_file_rows.append(row)
+            output_file_rows[-1].update({'status': ''})
+            # Creating/Updating Edx User, Edx User Profile and QVerse User Profile
+            # will be an atomic operation. If one operation fails the previous
+            # successfull operations will be reverted and no changes will be applied.
+            with transaction.atomic():
+                edx_user, _ = _create_or_update_edx_user(row)
+                _create_or_update_edx_user_profile(edx_user)
+                _create_or_update_qverse_user_profile(edx_user, row, output_file_rows)
+        except (IntegrityError, ValueError):
+            LOGGER.exception('Error while creating/updating {} user and its profiles.'.format(row.get('regno')))
+            output_file_rows[-1].update({'status': 'Failed'})
+        except Department.DoesNotExist:
+            LOGGER.exception('The Department with id {} does not exist.'.format(row.get('departmentid')))
+            output_file_rows[-1].update({'status': 'Failed'})
+    csv_file.close()
+    _write_status(instance.admission_file.path, output_file_rows)
+
+
+def _create_or_update_edx_user(user_info):
+    """
+    Creates/Updates edx user from given information.
+
+    Arguments:
+        user_info - A dict containing user information
+
+    Returns:
+        edx_user - (User) A newly created/updated User object
+        created - (bool) A flag indicating that user is created or not
+    """
+    user_data = {
+                    'password': make_password(user_info.get('regno')),
+                    'email': user_info.get('email'),
+                    'first_name': user_info.get('firstname'),
+                    'last_name': user_info.get('surname'),
+                    'is_active': True
+                }
+    edx_user, created = User.objects.get_or_create(username=user_info.get('regno'), defaults=user_data)
+    if created:
+        LOGGER.info('{} user has been created.'.format(edx_user.username))
+    else:
+        edx_user.first_name = user_info.get('firstname')
+        edx_user.last_name = user_info.get('surname')
+        edx_user.email = user_info.get('email')
+        edx_user.save()
+        LOGGER.info('{} user has been updated.'.format(edx_user.username))
+    return edx_user, created
```

Comment:
Why we are returning the created flag here if we don't it again.

Reply:
done.